File: reversebinary.java
import java.io.BufferedReader;
import java.io.InputStreamReader;

public class reversebinary{
    public static void main(String[] args){
        BufferedReader bi = new BufferedReader(new InputStreamReader(System.in)); // get the input

        try{
        	int N = Integer.parseInt(bi.readLine()); // gets the input N

            String originalBinary = Integer.toString(N, 2); // convert N to binary

            StringBuilder tmpBinary = new StringBuilder();
            tmpBinary.append(originalBinary);

            System.out.println(Integer.parseInt(tmpBinary.reverse().toString(),2)); //reverse the binary then convert back to decinmal using parseInt

        } 
		catch(Exception e){
            e.printStackTrace();
        }
    }
}

File: reversebinary.cpp
#include <iostream>
#include <string>  
#include <algorithm> // for reverse the sting
using namespace std;
int main(){
	int n;
	cin >> n;
	int d=0;
	string originalBinary = "";
	int r;
	// convert decimal to binary
	while(true){
		d = n/2;
		r = n%2;
		n = d;
		originalBinary += std::to_string(r);
		if(d == 1){
			originalBinary += std::to_string(d);
			break;
		}
	}
	// reverse the binary string
	//reverse(originalBinary.begin(), originalBinary.end());
	int num = stoi(originalBinary); // convert string back to integer
	
	int bin, rem, base = 1, dec = 0;
	while(num > 0){
		rem = num % 10;
		dec = dec + rem * base;
		base = base * 2;
		num = num / 10;
	}
	cout << dec << endl;
	return 0;
}
// this program works but it's too slow (time limit exceptions)
// runtime: exponential